Issue 14730 - Wrong closure var access with -inline

9rnsr opened the pull request:
https://issues.dlang.org/show_bug.cgi?id=14730

9rnsr committed:
fix Issue 14730 - Wrong closure var access with -inline

ibuclaw commented on test/runnable/closure.d:
```diff
@@ -898,6 +898,27 @@ void test12406()
+        struct S
+        {
+            int n;
+            int get() { return x; }     // x will be a cloaure variable
```
``closure``, not ``cloauasre  (I gave up trying to type the typo)

9rnsr replied:
Ouch. Fixed!

ibuclaw commented on src/func.c:
```diff
@@ -2145,6 +2145,104 @@ void FuncDeclaration::semantic3(Scope *sc)
+                xalign = v->alignment;
+            }
+            AggregateDeclaration::alignmember(xalign, memalignsize, &offset);
+            v->offset = offset;
```
I'm not sure what effect setting this in the frontend will have on GDC (I have a different way to build the closure struct).

9rnsr replied:
This code block determines the layout of heap-allocated closure object. If you have a better idea to make front-end code portable, I'll accept it.

ibuclaw replied:
As far as I can see it tries to establish the offset in the closure without actually associating them with the closure.

If I recall correctly, there are two places where offset is read on my side, generating user type layouts (not used for closure/frame generation), and when generating VarExp in code (should have already been converted to a closure field at that point).

So may be mostly harmless, but still unnerving as I might be using a slightly different layout than you in certain circumstances.  For instance with the presence of an out contract, all parameters are pushed into my closure on the basis that they *might* be used.

9rnsr replied:
I intend that this change won't introduce any difference in dmd on whether a variable will be actually be placed in closure. When a function local variable `offset` is set to nonzero here, it will be actually placed in closure. I believe the result is same with that the current dmd's glue-layer makes.

> For instance with the presence of an out contract, all parameters are pushed into my closure on the basis that they might be used.

The closure layout generated by GDC is sometimes different from the layout of dmd? If and only if GDC is actually using `VarDeclaration::offset` for the custom closure layout, you might need to mask this offset calculation code.

9rnsr committed:
Move back the closure vars offset calculation to glue layer.

It's necessary to avoid breaking of test14730x(). But I think it's questionable.
See: https://issues.dlang.org/show_bug.cgi?id=14771

9rnsr commented:
I noticed that, moving the closure vars offset calculation into front-end will cause breaking of currently working code. Look at the added test case `test14730x()`.

But I think it's questionable behavior, because I also found a @nogc violation issue around that.
https://issues.dlang.org/show_bug.cgi?id=14771

Anyway, I changed the bugfix way to more conservative.

9rnsr commented:
Rebased on master, and ready for ddmd.

ibuclaw commented:
OK.

ibuclaw commented:
Auto-merge toggled on
